sdk/src/metrics/instrument_metadata_validator.cc

```diff
@@ -0,0 +1,78 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef ENABLE_METRICS_PREVIEW
+#  include "opentelemetry/sdk/metrics/instrument_metadata_validator.h"
+#  include "opentelemetry/common/macros.h"
+#  include "opentelemetry/nostd/string_view.h"
+
+#  include <algorithm>
+#  include <iostream>
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace sdk
+{
+namespace metrics
+{
+const size_t kMaxSize = 63;
+// instrument-name = ALPHA 0*62 ("_" / "." / "-" / ALPHA / DIGIT)
+const std::string kInstrumentNamePattern = "[a-zA-Z][-_.a-zA-Z0-9]{0,62}";
+//
+const std::string kInstrumentUnitPattern = "[\x01-\x7F]{0,63}";
+// instrument-unit = It can have a maximum length of 63 ASCII chars
+
+InstrumentMetaDataValidator::InstrumentMetaDataValidator()
+#  if HAVE_WORKING_REGEX
+    : name_reg_key_{kInstrumentNamePattern}, unit_reg_key_
+{
+  kInstrumentUnitPattern
+}
+#  endif
+{}
+
+bool InstrumentMetaDataValidator::ValidateName(nostd::string_view name) const
+{
+#  if HAVE_WORKING_REGEX
+  return std::regex_match(name.data(), name_reg_key_);
+#  else
+  // size atmost 63 chars
+  if (name.size() > kMaxSize)
+  {
+    return false;
+  }
+  // first char should be alphanumeric
```

Comment:
First char should be alpha, not alphanumeric?

Reply:
yeah good point. fixed the comment.